messaging/src/main/java/org/axonframework/messaging/MessageStream.java

```diff
@@ -248,7 +357,41 @@ default MessageStream<M> concatWith(@NotNull MessageStream<M> other) {
      * @param completeHandler The {@link Runnable} to invoke when the {@link MessageStream stream} completes normally.
      * @return A {@link MessageStream stream} that invokes the {@code completeHandler} upon normal completion.
      */
-    default MessageStream<M> whenComplete(@NotNull Runnable completeHandler) {
+    default MessageStream<M> whenComplete(@Nonnull Runnable completeHandler) {
         return new CompletionCallbackMessageStream<>(this, completeHandler);
     }
+
+    /**
+     * A {@link MessageStream}-specific container of {@link Message} implementations.
+     * <p>
+     * May be implemented to support {@link Entry entries} that contain several objects. As such, this interface may be
+     * regarded as a tuple.
+     *
+     * @param <M> The type of {@link Message} contained in this {@link Entry} implementation.
+     * @author Allard Buijze
+     * @author Milan Savić
+     * @author Mitchell Herrijgers
+     * @author Steven van Beelen
+     * @since 5.0.0
+     */
+    interface Entry<M extends Message<?>> extends Context {
```

Comment:
I don't think we really settled on a name as a team. I personally settled on `Entry` because of the relationship to `Map.Entry`. To me, `Map.Entry` makes sense, but is similarly not extremely descriptive about what's there. I figured `MessageStream.Entry` is similar. Although, granted, the `Entry` imposes the existence of a `Message`. Then again, the fact this is an inner class to the `MessageStream` would make that apparent.

Regardless, let's discuss this point in our meeting tomorrow too :+1: